### ISSUE TITLE
Improve performance of Aggregator

### ISSUE THREAD

**Issue by dsaccon** [bmoscon/cryptostore#109]
** Problem **
I am running a configuration with roughly 2000 instrument pairs across a dozen or so exchanges, collecting TRADES data type. Storage is InfluxDB and cache is Redis. storage_interval as per my config file is 10s

I noticed that it was common for the loop in Aggregator that is cycling through these exchanges + pairs to not complete before the 10s is elapsed and thus I would see messages generated from the below line from aggregator.py

LOG.warning("Storage operations currently take %.1f seconds, longer than the interval of %d", total, interval)

This caused a number of issues in my deployment

** Possible solution **
I was able to address this in my environment by collating all data queried from cache in the loop to a list, and then dumping the contents of that list in bulk after the loop completed. This sped things up by about an order of magnitude (15-20s before, less than 2s after)

To make this work I had to also adjust the write() function in InfluxDB class from influx.py. Currently this method takes in *exchange* and *pair* args which are used to create the Influx line protocol string. However this same information is available in self.data so I just used that. This allowed me to batch up trades from all exchanges/pairs in the Aggregator loop and put into one write() transaction

I haven't looked at other storage backends so can't say if the same could also be done there

What I am doing also relies on there being only one data type per bulk write. Looking at the influx.py code, it looks like that could maybe be made generic so it is possible to mix data types

I thought changes like these could maybe help others in the community wanting to run scaled configurations

**Comment by bmoscon** [comment]
thats a ton of data, so this isnt that unexpected. I'm working on making the process that takes data from cache to storage run in parallel, so that would solve this issue, assuming your redis instance and influx instance could handle N workers reading/writing simultaneously

**Comment by yohplala** [comment]
@bmoscon 
Bryant, reading you, do you have some pointers to configure redis so that it can handle n workers?
(a tuto?)
At the moment,  i have been using 2 instances of cryptostore, each one with its own port for redis, and each having a different redis service (systemd). So this is not the same case.

**Comment by bmoscon** [comment]
it can handle thousands of clients by default, you don't need to configure anything for this.

**Comment by yohplala** [comment]
Hi @sirdavealot,
Sorry to disturb you.
Please, would you mind giving a test to this develop branch, that aims for improvement of the aggregator?
```
git clone -b redis_read_optimization https://github.com/yohplala/cryptostore
cd cryptostore
python3 setup.py install develop
```
And then run your usual config with Influx.

It will not reach the 2s you are targetting, but results should be closer to 12-14s (hopefully)
Most notably peak memory should be reduced but I don't know if you have any way to measure before/after.
I don't have Influx, and it would be nice to have confirmation the PR is ok with Influx.
Thanks for your help!

**Comment by sirdavealot** [comment]
hi @yohplala - sorry for the delay. Happy to help but note the environment I was running for #109 was slightly different (collecting trades only, no order book data). Running order book for my original config would overwhelm my local machine, or even a powerful machine in the cloud probably. Sounds like this doesn't matter, you just need to confirm it works with influx, right?

Also I saw a comment from @bmoscon in #113 about testing this so not sure where you are at, let me know if you still need help

Btw, influxdb is just another docker container service so I'm curious why you haven't been able to run it?

Thanks

**Comment by sirdavealot** [comment]
@yohplala @bmoscon - in addition to #113 there could probably also be speedups from using the redis pipelining feature. I have had success using it in the past

Considering the sequential calls to read() that happen in the loop in aggregator.py, the idea would be to batch these up using redis pipelines similar to how I batched up the Influx writes in #109. I didn't worry about implementing this at the time because it was the influx writes which dominated the delays I was seeing

Its a little different than the solution of using parallel workers but with the advantage of not requiring additional system resources (CPU cores). But probably I don't have the full context with all the work going on with cryptostore. Happy to look into it further if it could be helpful to the project

**Comment by yohplala** [comment]
> hi @yohplala - sorry for the delay. Happy to help but note the environment I was running for #109 was slightly different (collecting trades only, no order book data). Running order book for my original config would overwhelm my local machine, or even a powerful machine in the cloud probably. Sounds like this doesn't matter, you just need to confirm it works with influx, right?

Hi @sirdavealot,
Indeed it is just confirmation that it works.
You can select a single pair, use `l2_order_book`, and `trades`. As said speed-up should be more noticable for `l2_order_book`, than `trades`.

> Also I saw a comment from @bmoscon in #113 about testing this so not sure where you are at, let me know if you still need help

I think it is better to have more tests than not enough. :) If for you it is not a big trouble to set this test up, that would be great.
The ideal thing would be to test with current cryptostore, and with the branch I gave you the link, to report perf. improvements.

> Btw, influxdb is just another docker container service so I'm curious why you haven't been able to run it?

Because I am a newbie with Influx (never took the timee to give it a look) and also with docker (though I did give it some time, but gave up after unsuccessful trials.

> @yohplala @bmoscon - in addition to #113 there could probably also be speedups from using the redis pipelining feature. I have had success using it in the past. Considering the sequential calls to read() that happen in the loop in aggregator.py, the idea would be to batch these up using redis pipelines similar to how I batched up the Influx writes in #109. I didn't worry about implementing this at the time because it was the influx writes which dominated the delays I was seeing.

I guess any idea is welcome.
Does pipelines in redis mean parallel data streams?
If so, I am guessing understand that peak memory will be higher right?
Actually, peak memory is my trouble, more than CPU. It is with memory that I had trouble, which is why in the commit I am proposing, for `order_book`, I see generator as being a quite efficient solution.
If you have a way to monitor the peak memory with your tests, you will see that peak memory ought to be drastically reduced. This ought to be especially true in the case of Influx.

Thanks for your help!
Bests

**Comment by sirdavealot** [comment]
@yohplala - I tested and it looks like writes to Influx work with your changes. Can't comment on perf improvement or memory though, only ran with your version. Note, I had to edit the code to reduce the number of writes. Just BITMEX-XBTUSD l2 book had millions of entries for the first storage write when running the script. First time I have used cryptostore to store order book updates (previously only used for trades) so not sure if I'm missing something here

re Pipelining - it isn't a concurrency or parallelism scheme, its just a way to batch multiple redis operations together. The batch rides on the same TCP connection (reduce handshake round trips) and reducing syscalls in redis when it serves the requests. But pipelines will probably require extra memory for redis. If all these services are being run on the same host, hard to say what the net effect on system resources is

**Comment by yohplala** [comment]
> @yohplala - I tested and it looks like writes to Influx work with your changes. Can't comment on perf improvement or memory though, only ran with your version. Note, I had to edit the code to reduce the number of writes. Just BITMEX-XBTUSD l2 book had millions of entries for the first storage write when running the script. First time I have used cryptostore to store order book updates (previously only used for trades) so not sure if I'm missing something here
> 
> re Pipelining - it isn't a concurrency or parallelism scheme, its just a way to batch multiple redis operations together. The batch rides on the same TCP connection (reduce handshake round trips) and reducing syscalls in redis when it serves the requests. But pipelines will probably require extra memory for redis. If all these services are being run on the same host, hard to say what the net effect on system resources is

@sirdavealot, a lot of thanks!
I didn't expect any trouble as the input data to influx.write was double-checked already through other tests I ran, and in influx.write(), I could only see `self.data` being used once  in a `for` loop, but good to have confirmation 👍 
Thanks again for the confirmation.

Regarding pipeline, ok, good to know.
Myself, I am sorry, I am prioritizing other devs at the moment so I won't have a look into this direction, at least not before several months... 

Thanks again!


PS: you could have tried another exchange like `kraken` to have fewer lines per order book.
The 1st order book retrieved is a snapshot, so it contains ALL lines of the order book, whatever `storage_interval`
I am guessing BITMEX order book is probably one of the biggest, along with BINANCE and COINBASE.

Thanks again!

**Comment by sirdavealot** [comment]
@yohplala - You are right, it was my config settings, messages are reduced now. Thanks for the suggestion

Maybe when I get a chance I'll look at pipelining implementation. But at least from my original use case it looked like writing to storage was a bigger bottleneck in this part of the system.. Is your change for #113 planned to get pulled in btw?

**Comment by yohplala** [comment]
> Is your change for #113 planned to get pulled in btw?

I hope so. It is already in production since a week on my side. :)

Hi @bmoscon,
Please, let us know if you need anymore details/tests greading PR #113 
Thanks

**Comment by bmoscon** [comment]
113 is in - i think this will fix many of your issues, if so, let me know and we can close this

**Comment by sirdavealot** [comment]
> 113 is in - i think this will fix many of your issues, if so, let me know and we can close this

@bmoscon you mean the #113 fix will address #109? Looking at the commits I don't think it addresses the main perf issues since #109 is about how the loop in aggregator.py works and this file wasn't touched in #113 

Might be best if I test the updated code to confirm and then show you the changes I made on a separate branch

**Comment by bmoscon** [comment]
i see, i misunderstood the issue. I think the real fix here is to parallelize the writes.
